### PR TITLE
Get Epoch (&Boundary) Attestations Helper Functions

### PR DESCRIPTION
**Pull request by terenc3t**
This PR implements the following helper functions from the [spec](https://github.com/ethereum/eth2.0-specs/commit/c43724132b06086ae51a11001794dce802c74644). They are needed for epoch processing.

```python
this_epoch_attestations = [a for a in state.latest_attestations if state.slot - EPOCH_LENGTH <= a.data.slot < state.slot]
this_epoch_boundary_attestations = [a for a in this_epoch_attestations if a.data.epoch_boundary_root == get_block_root(state, state.slot-EPOCH_LENGTH) and a.justified_slot == state.justified_slot
```

### CODE REVIEW

**Comment by rauljordan** [comment]
nitpick but can we move this stuff to `core/epoch` and remove the `Epoch` initial prefix from the funcs so they can be imported and called like
```go
epoch.BoundaryAttestations(...)
```

File: beacon-chain/core/state/epoch_processing.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	for _, attestation := range state.LatestAttestations {
+
+		// If the state slot is less than epochLength, then the earliestSlot would
+		// result in a negative number. Therefore we should default to
+		// earliestSlot = 0 in this case.
+		if state.Slot > epochLength {
+			earliestSlot = state.Slot - epochLength
+		}
+
+		if earliestSlot <= attestation.Data.Slot && attestation.Data.Slot < state.Slot {
```
Comment (rauljordan) [inline beacon-chain/core/state/epoch_processing.go]: Use the getters here when nesting dot properties of proto structs. For example, prefer `attestation.GetData().GetSlot()` as that will avoid any panics if data is undefined. In the future, we can replace this with our own `*types.Attestation` once we complete the refactor.

File: beacon-chain/core/state/epoch_processing.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+// EpochBoundaryAttestations returns the pending attestations from
+// the epoch boundary block.
+//
+// Spec pseudocode definition:
+//   [a for a in this_epoch_attestations if a.data.epoch_boundary_root ==
+//   get_block_root(state, state.slot-EPOCH_LENGTH) and a.justified_slot ==
+//   if state.slot - EPOCH_LENGTH <= a.data.slot < state.slot]
+func EpochBoundaryAttestations(state *pb.BeaconState, thisEpochAttestations []*pb.PendingAttestationRecord) ([]*pb.PendingAttestationRecord, error) {
```
Comment (rauljordan) [inline beacon-chain/core/state/epoch_processing.go]: line too long

File: beacon-chain/core/state/epoch_processing_test.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	"bytes"
+	"testing"
+
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+func TestEpochAttestations_ok(t *testing.T) {
+	if params.BeaconConfig().EpochLength != 64 {
+		t.Errorf("EpochLength should be 64 for these tests to pass")
```
Comment (rauljordan) [inline beacon-chain/core/state/epoch_processing_test.go]: This should fatal

File: beacon-chain/core/epoch/epoch_processing.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	}
+	return thisEpochAttestations
+}
+
+// BoundaryAttestations returns the pending attestations from
+// the epoch boundary block.
+//
+// Spec pseudocode definition:
+//   [a for a in this_epoch_attestations if a.data.epoch_boundary_root ==
+//   get_block_root(state, state.slot-EPOCH_LENGTH) and a.justified_slot ==
```
Comment (prestonvanloon) [inline beacon-chain/core/epoch/epoch_processing.go]: Typo?

File: beacon-chain/core/epoch/epoch_processing.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+//   if state.slot - EPOCH_LENGTH <= a.data.slot < state.slot]
```
Comment (prestonvanloon) [inline beacon-chain/core/epoch/epoch_processing.go]: What is this? I don’t see this logic anywhere

File: beacon-chain/core/state/epoch_processing_test.go (hunk 3 above)
Comment (prestonvanloon) [reply beacon-chain/core/state/epoch_processing_test.go]: I disagree, if the test continues and passes then you have even bigger problems. Change to Errorf please

**Pull request merged**